### PR TITLE
[MISC] Create data-models version 1

### PR DESCRIPTION
**Pull request by sinclert-canonical**
This PR creates version 1 of those libraries that would suffer breaking changes due to the support of Pydantic 2:
- `data_models`.
- ~`upgrade`~.

Given that this is the first PR that bumps the major version of some of the libraries, some code shuffling was required to give the repository a cohesive view of what is still in `v0`, and what is available both in `v0` and `v1`.

---

### Pydantic 2 specific changes:
Pydantic 2 breaking changes can be seen in commit https://github.com/canonical/data-platform-libs/pull/201/commits/4fbfe4f8877340dc0f06ce72fc596a106ad8e547.

<details>
  <summary>Code substitutions</summary>

  ```diff
  - from pydantic import validator
  + from pydantic import field_validator

  - from pydantic import root_validator
  + from pydantic import model_validator

  - field.outer_type_
  + field.annotation

  - obj.dict
  + obj.model_dump

  - obj.__fields__
  + obj.model_fields

  - type(self.dependency_model)(**json.loads(deps))
  + type(self.dependency_model).model_validate_json(deps)

  - json.dumps(self.dependency_model.dict())
  + self.dependency_model.model_dump_json()
  ```

</details>

---

### Code shuffling changes:

- Moved + adapted all previous integration tests into [tests/v0/integration](https://github.com/canonical/data-platform-libs/tree/sinclert/pydantic-2/tests/v0/integration).
- Moved + adapted all previous unit tests into [tests/v0/unit](https://github.com/canonical/data-platform-libs/tree/sinclert/pydantic-2/tests/v0/unit).
- Copied + adapted `data_models.py` ~and `upgrade.py`~ tests into [tests/v1/unit](https://github.com/canonical/data-platform-libs/tree/sinclert/pydantic-2/tests/v1/unit).
- Copied + adapted `data_models.py` ~and `upgrade.py`~ lib into [libs/charms/data_platform_libs/v1](https://github.com/canonical/data-platform-libs/tree/sinclert/pydantic-2/lib/charms/data_platform_libs/v1).
- Adapted Tox.ini environments to allow for `-v0` and `-v1` suffixes.
- Adapted CI workflow to use the new Tox environments.

### CODE REVIEW

**Comment by sinclert-canonical** [comment]
The major version bump is proposed so backwards compatibility is not broken when client charms try to inject a _dependency model_ based on Pydantic 1 `BaseModel` class, within the upgrade library ([reference](https://github.com/canonical/data-platform-libs/blob/c5696736e6582b8e93794f6fe15f356ac620efa0/lib/charms/data_platform_libs/v0/upgrade.py#L509)).

**Comment by sinclert-canonical** [comment]
Depends on PR https://github.com/canonical/data-platform-libs/pull/214.

**Comment by sinclert-canonical** [comment]
Depends on PR https://github.com/canonical/data-platform-libs/pull/214.

**Comment by dragomirp** [comment]
Tested on https://github.com/canonical/postgresql-operator/pull/902 for PG.

**Review by dragomirp — APPROVED** [review]

**Review by taurus-forever — APPROVED** [review]
Thank you @sinclert-canonical !!!

File: tests/v0/integration/application-s3-charm/charmcraft.yaml [hunk 1]
```diff
@@ -18,3 +18,7 @@ bases:
     run-on:
       - name: "ubuntu"
         channel: "20.04"
+parts:
+  charm:
+    charm-binary-python-packages:
+      - opentelemetry-api
```
[the remark this answers is not in the document]
Comment (sinclert-canonical) [reply tests/v0/integration/application-s3-charm/charmcraft.yaml:24]: Commits https://github.com/canonical/data-platform-libs/pull/201/commits/76ed1e64967b7c90a1a3a3df19f281e1d5416895 and https://github.com/canonical/data-platform-libs/pull/201/commits/11481af92d50627e182f5a117c85f0062c372faa (the one containing the referenced lines) have been extracted from PR https://github.com/canonical/data-platform-libs/pull/214, a pre-requisite for having this one merged. I only put them here so you all can see the satisfying "✅" on CI status, and see that none of the proposed code shuffling breaks existing tests.

In summary: those changes should not be part of this PR, but part of https://github.com/canonical/data-platform-libs/pull/214.

File: tests/v0/integration/conftest.py [hunk 2]
```diff
@@ -48,72 +48,72 @@ async def build_charm(charm_path, bases_index: int = None) -> Path:
 def copy_data_interfaces_library_into_charm(ops_test: OpsTest):
     """Copy the data_interfaces library to the different charm folder."""
     library_path = "lib/charms/data_platform_libs/v0/data_interfaces.py"
-    install_path = "tests/integration/database-charm/" + library_path
+    install_path = "tests/v0/integration/database-charm/" + library_path
```
Comment (deusebio) [inline tests/v0/integration/conftest.py:51]: **suggestion** maybe one could rather use relative path and infer the root path (if needed) within the script

**Review by deusebio — APPROVED** [review]
Looks good and I don't have strong concerns. I could see a few improvements to structure better the various versions, with a parametric behaviour of `v0` and `v1`, like:

* Use relative path without `v*` and autoinfer the absolute path/root for the tests
* Use an env `VERSION` in the general env tox file, such that the various environment can be used for both `v1` and `v0` making it easily extensible. See [here](https://github.com/canonical/katib-operators/blob/main/tox.ini) for an example on how to combine it (see the charms - controller, manager, ui - combined with the linting, unit, integration), instead of defining the single cartersian combination
* what about moving the requirements into the lib folder? I'm not sure, but my gut feelings tells me that having requirements-v*.txt at the root is not great. Ideally, we should have a vN folder in source, and a vN folder in the tests, and also the dependendecies would be there. Otherwise, another pattern I would prefer is also using a folder `requirements/vN/requirements.txt` if reqs in the source is not a thing

Anyhow, all these comments are suggestions, nipticks and minor preferences. Up to you or the other reviewers to see if there is something worth following up on. These are not critical, so I would also be fine to merge the PR also as-is for my part (hence approving).

File: tests/v0/integration/application-s3-charm/charmcraft.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (sinclert-canonical) [reply tests/v0/integration/application-s3-charm/charmcraft.yaml]: Rebased from `main` branch.

File: tests/v0/integration/conftest.py (hunk 2 above)
Comment (sinclert-canonical) [reply tests/v0/integration/conftest.py:51]: I am afraid the complexity introduced by the `pathlib` logic to automatically infer whether we are in V0 or V1 of the integration tests, would be higher than the complexity of stating `v{0,1,...}` on the specific integration test we are modifying.

Open to the improvement if people feel strongly about it. Otherwise, I would rather keep it dumb and explicit.

**Comment by deusebio** [comment]
Cool! Thanks a lot for following up on the comments @sinclert-canonical !! I had already approved, and now I would just approve MORE :joy:

File: lib/charms/data_platform_libs/v1/data_models.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
+# Copyright 2024 Canonical Ltd.
```
Comment (Gu1nness) [inline lib/charms/data_platform_libs/v1/data_models.py:1]: nit: `# Copyright 2025 Canonical Ltd.`
to be updated in the future

**Review by Gu1nness — APPROVED** [review]
